text/0000-cargo-features2.md

```diff
@@ -0,0 +1,520 @@
+```
+
+For packages that encounter a problem due to missing feature declarations, it
+is backwards-compatible to add the missing features. Adding those missing
+features should not affect projects using the old resolver.
+
+It is intended that `resolver = "2"` will likely become the default setting in
+a future Rust Edition. See ["Default opt-in"](#default-opt-in) below for more
+details.
+
+## New command-line behavior
+
+The following changes are made to the behavior of selecting features on the
+command-line.
+
+* Features listed in the `--features` flag no longer pay attention to the
+  package in the current directory. Instead, it only enables the given
+  features for the selected packages. Additionally, the features are enabled
+  only if the the package defines the given features.
+
+  For example:
+
+      cargo build -p member1 -p member2 --features foo,bar
+
+  In this situation, features "foo" and "bar" are enabled on the given members
+  only if the member defines that feature. It is still an error if none of the
+  selected packages defines a given feature.
+
+* Features for individual packages can be enabled by using
+  `member_name/feature_name` syntax. For example, `cargo build --workspace
+  --feature member_name/feature_name` will build all packages in a workspace,
+  and enable the given feature only for the given member.
+
+* The `--features` and `--no-default-features` flags may now be used in the
+  root of a virtual workspace.
+
+The ability to set features for non-workspace members is not allowed, as the
+resolver fundamentally does not support that ability.
+
+The first change is only enabled if the `resolver = "2"` value is set in the
+workspace manifest because it is a backwards-incompatible change. The other
+changes are intended to be stabilized for everyone, as they only extend
+previously invalid usage.
+
+## `cargo metadata`
+
+At this time, the `cargo metadata` command will not be changed to expose the
+new feature resolver. The "features" field will continue to display the
+features as computed by the original dependency resolver.
+
+Properly expressing the dependency graph with features would require a number
+of changes to `cargo metadata` that can add complexity to the interface. For
+example, the following flags would need to be added to properly show how
+features are selected:
+
+* Workspace selection flags (`-p`, `--workspace`, `--exclude`).
+* Whether or not dev-dependencies are included (`--dep-kinds`?).
+
+Additionally, the current graph structure does not expose the host-vs-target
+dependency relationship, among other issues.
+
+It is intended that this will be addressed at some point in the future.
+Feedback on desired use cases for feature information will help define the
+solution. A possible alternative is to stabilize the [`--unit-graph`] flag,
+which exposes Cargo's internal graph structure, which accurately indicates the
+actual dependency relationships and uses the new feature resolver.
```

Comment:
I am not sure how `unit-graph`'s output looks like, but I want to say that `cargo metadata` and https://lib.rs/crates/cargo_metadata is used by a lot of projects in the ecosystem. I reckon we need to look into fixing this when this new resolver is implemented.

IIRC even rust-analyzer depends on it to understand the dependency graph. /cc @matklad